### ISSUE TITLE
RFC: JSON Protocol v4 Schema 

### ISSUE THREAD

**Issue by IanLondon** [Opentrons/opentrons#4836]
## What's new: overview

1. Top-level `"$otSharedSchema": "#/protocol/4"` to identify what kind of Opentrons JSON file we're dealing with (see latest modules schema + https://github.com/Opentrons/opentrons/issues/4601 for explanation)
2. Top-level `modules` key (acts just like `labware` or `pipettes` to describe what modules are in the protocol and what location they're at)
3. Additional commands to support magnetic, temperature, and thermocycler modules (detailed below)

## Background: PD's requirements for non-blocking versions of commands

PD's product requirements include the need to pipette and maybe do other module commands while a module is running a time-consuming step. Specifically:
* User should be able to do any pipetting they want while the temperature deck / thermocycler is moving towards a target temp.
* User should be able to specify "wait until target temp: ___C" to gate further protocol execution until the condition is reached.
* The thermocycler should be able to run a multi-part "profile" while non-thermocycler pipetting steps happen on the robot. Finally when the user wants to interact with the thermocycler again, they'll need to put a "wait until thermocycler profile complete" step in before PD allows it.

In this respect, PD needs an executor that is more powerful than our current user-facing PAPI. We'll have to make non-blocking versions of some module methods to accomplish this in the executor.

## Linked PR

The PR for the proposed schema is #4846. JSON schema isn't the most readable thing, so before you dive into that, please read the summary below!

## New commands: SUMMARY

Please note the below is just a summary to help outline what's being proposed before you dive into the Flow def / JSON schema in the linked PR that contains all the details!

### Magnetic module

`magneticModule/disengageMagnet`: disengages magnet (to home position). Blocking.

`magneticModule/engageMagnet`: engaged magnet to specified height (mm above labware Z-bottom). Blocking.

* takes an `engageHeight` parameter (mm above labware Z-bottom)

### Temperature module

`temperatureModule/setTargetTemperature`: sets a target temperature for the temperature module to approach/maintain, does NOT block further execution of the protocol

* takes a `temperature` param

`temperatureModule/awaitTemperature`: block further protocol execution until temperature module reaches specified temperature. (PS: this should NOT block thermocycler profile execution, which is on a parallel timeline)

* takes a `temperature` param

`temperatureModule/deactivate`: temperature module stops active heating or cooling, just passively sits there (instantaneous, nothing to await)

### Thermocycler

`thermocycler/setTargetBlockTemperature` + `thermocycler/setTargetLidTemperature`: non-blocking command to set target temperature, similar to temp module setTargetTemperature command. But TC has both block and lid temps to deal with!

`thermocycler/awaitBlockTemperature` + `thermocycler/awaitLidTemperature`: similar to temp module's awaitTemperature command.

`thermocycler/deactivateBlock` + `thermocycler/deactivateLid`: similar to temp module's deactivate command.

`thermocycler/openLid` + `thermocycler/closeLid`: Blocking. Open or close the lid. (Aside: PAPI should throw an error if you try to interact with the thermocycler with a pipette when the lid is closed. This behavior is already in PAPI :+1:)

`thermocycler/runProfile`: immediately begin running specified profile. You should probably close the lid before doing this. The profile runs on a parallel timeline outside of normal protocol command execution flow - so other (non-thermocycler) commands can execute while the profile is running. Probably any thermocycler commands following this should throw an error, unless there's a `thermocycler/awaitProfileComplete` command to make sure the profile is finished.

- The "profile" parameter can be specified like: `Array<{| temperature: number, holdTime: number |}>`

`thermocycler/awaitProfileComplete`: block protocol execution until the current thermocycler profile is complete (if there is no profile running, just move along and maybe log a warning?)

## Q & A

**What about the liquid volume parameters for temp module + thermocycler?**

Great question, I'm not sure what commands need them and what commands don't! Or should "thermocycler/setLiquidVolume" be its own command? Are ramp rates necessary for JSON commands? Please give some feedback here!

Also if you notice any other missing parameters from commands (please see the linked PR, I don't give all the details in the summary above), please comment!

-----

**Why distinguish commands by module type?** Instead of having `temperatureModule/setTargetTemperature` and `thermocycler/setTarget[lid|block]Temperature`, we could just have a single `setTargetTemperature` command. It's good to keep the number of commands we have manageably low. It's a bit redundant to specify the `module` ID in the `params` and then also specify the module's type in the `command` itself.

However, the commands have meanings that are specific to a particular module. Even "set target temperature" means something different for temperature module (there's only body temp) vs thermocycler (either body or lid could be specified). If we were to share the same command across different module types, we might end up making `params` overly complex to accomodate different information.

Hopefully this division between module types will keep our commands and the shape of their `params` more robust over time.

Credit to Seth for coming up with this idea!

---

**Why have additional commands for `thermocycler/runProfile` and
`thermocycler/awaitProfileComplete`?** To accomplish running a thermocycler profile, you don't strictly need these new commands. Instead, you could do:

```
setTargetBlockTemperature 80
awaitBlockTemperature 80
delay X time

setTargetBlockTemperature 12
awaitBlockTemperature 12
delay X time

// ... keep going until the profile is finished
```

However, in our existing designs for upcoming thermocycler work in PD, the user can run a thermocycler profile in a non-blocking manner. (See https://github.com/Opentrons/opentrons/issues/4655). This isn't quite possible with the above commands, because you don't have a way of moving to the new profile step without delaying the whole protocol. So instead, the proposed commands allow a protocol to do:

```
runProfile [80C for X time, 12C for X time, ...]
// ... pipetting stuff here that doesn't involve thermocycler
awaitProfileComplete
openLid
// ... pipette stuff out of the thermocycler into tubes etc
```

This manages asynchrony the way we need to meet the product requirements.

---

**Still, this is a lot of new commands.** Could we combine *some* of them?

* For thermocycler we combine the lid vs body commands by a param like `lidOrBody: 'lid' | 'body'`?
* Could we make `openLid` / `closeLid` into "moveLid" and likewise distinguish them with params?

I'm open to these suggestions. I think splitting them makes it slightly simpler, but I don't have a strong objection to combining some of them as long as it doesn't make the `params` too polymorphic. 

---

**These commands have ordering restrictions**, for example, you need to set a temperature before you can await it. How do we deal with nonsensical command ordering?

For decent UX, this needs to be dealt with in PD first. It would be bad to have PD save a protocol that looks fine, and then upon upload, the user gets an error like "Invalid protocol commands. Awaiting a temperature that will never be reached."

The simple version of this problem that we have successfully dealt with already is tip state. PD is responsible for making protocols that pick up a tip before the pipette aspirates. PAPI will throw an error if you try to `pipette.aspirate(...)` without a tip. But PD hasn't run into this because it guarantees on its own end that it will manage tips sanely. We'll have more work on the PD end to make similar sort of guarantees about the modules commands that PD puts out.

It's an open question about whether or not we want PAPI to extensively report these sorts of errors about awaiting things that won't happen. If it does slip past PD somehow, it's a better UX to be blocked by an error than to run a protocol that never finishes. But it's still a PD bug. IMO it would be a very nice to have but wouldn't block implementation of the v4 executor if it comes down to making it into an upcoming release or not.
